### ISSUE TITLE
trace: support string interpolation

### ISSUE THREAD

**Issue by Ralith** [tokio-rs/tracing#83]
Log readability is significantly improved by interleaving information with text, rather than supplying a static string followed by a set of key-value pairs. tokio-trace's macros should support constructing such interpolated strings while preserving the interpolated data as structured fields.

Additionally, any out-of-the-box subscribers targeting direct human consumption (such as by printing to stdout) should not duplicate interpolated values in its output. This could be accomplished without compromising support for users that prefer not to use interpolation by using a distinct field name for message strings that include interpolated data, and establishing that human-oriented subscribers should only display unrecognized key-value pairs if an interpolated message field is not present.

**Comment by samschlegel** [comment]
:wave: Dropping a line here in reference to a convo in the #tracing channel in Discord.

Currently we use `slog` in a bunch of our libs/services which has a syntax like

```rust
info!(logger, "msg: {}", 1, formatted_field="a"; "additional_field" => 4)
```

The fields before the ; are used for the format string, and the result of that is used as `msg` in the json output. The fields after are just added to the json directly.

I'm not actually a huge fan of this, mostly cause it leads to duplication of fields, and the syntax differs between then in a weird wait. Something that might be nice could be if the syntax was basically just like `format!/println!` but it didn't error when named arguments weren't used, and all named arguments were added as fields, but I haven't totally thought through the implications of that. Like I'm not sure how `field = ?value` would work if the field was used multiple times.

**Comment by Ralith** [comment]
Since Rust [RFC 2795](https://rust-lang.github.io/rfcs/2795-format-args-implicit-identifiers.html) will be stabilized in 1.58, the ergonomic semantics outlined by @samschlegel above should soon be achievable. All we'll need to do is modify `event!` and `span!` to interpret format arguments as fields, and use implicit format argument captures for them so that unused fields do not become an error.

**Comment by CryoMyst** [comment]
Is there anything currently blocking this feature from happening?

**Comment by Ralith** [comment]
No, it's fairly straightforwardly doable now, just nobody's had time to get to it.
